provisioning_client/adapters/hsm_client_riot.c

```diff
@@ -780,7 +786,7 @@ char* hsm_client_riot_create_leaf_cert(HSM_CLIENT_HANDLE handle, const char* com
         LEAF_CERT_TBS_DATA.SubjectCommon = common_name;
 
         DERInitContext(&leaf_ctx, leaf_buffer, DER_MAX_TBS);
-        if (X509GetDeviceCertTBS(&leaf_ctx, &LEAF_CERT_TBS_DATA, &leaf_id_pub) != 0)
+        if (X509GetDeviceCertTBS(&leaf_ctx, &LEAF_CERT_TBS_DATA, &leaf_id_pub, (uint8_t*)&riot_info->ca_root_pub, sizeof(riot_info->ca_root_pub)) != 0)
```

Comment:
Why does the TBS for leaf require the root cert? Is this building the chain or just the EKU property?

Reply:
Whoops, copy paste error, I believe that should be leaf_id_pub

Reply:
I'm going to assume from the context here that the term "leaf cert" is meant to imply "Alias cert"?  The X509GetDeviceCertTBS function will create the to be signed region for a DeviceID certificate.  This is not a general purpose function.  In construction of the Alias cert you must instead call X509GetAliasCertTBS.  

There are x509 extensions that are included in an Alias cert that are not included in the DeviceID cert and vice versa.

Reply:
This is actually creating an end-entity certificate, there are helper functions that create alias cert and create alias cert key pair, however the call in question is an actual production API that creates a leaf cert:
```C
char* hsm_client_riot_create_leaf_cert(HSM_CLIENT_HANDLE handle, const char* common_name)
{
    char* result;

    // The static data fields that make up the DeviceID Cert "to be signed" region
    RIOT_X509_TBS_DATA LEAF_CERT_TBS_DATA = {
        { 0x5E, 0x4D, 0x3C, 0x2B, 0x1A }, RIOT_CA_CERT_NAME, "MSR_TEST", "US",
        "170101000000Z", "370101000000Z", "", "MSR_TEST", "US" };

    if (handle == NULL || common_name == NULL)
    {
        /* Codes_SRS_HSM_CLIENT_RIOT_07_030: [ If handle or common_name is NULL, hsm_client_riot_create_leaf_cert shall return NULL. ] */
        LogError("invalid parameter specified.");
        result = NULL;
    }
    else
    {
        RIOT_STATUS status;
        uint8_t leaf_buffer[DER_MAX_TBS] = { 0 };
        DERBuilderContext leaf_ctx = { 0 };
        RIOT_ECC_PUBLIC     leaf_id_pub;
        RIOT_ECC_SIGNATURE tbs_sig = { 0 };

        HSM_CLIENT_X509_INFO* riot_info = (HSM_CLIENT_X509_INFO*)handle;

        LEAF_CERT_TBS_DATA.SubjectCommon = common_name;

        DERInitContext(&leaf_ctx, leaf_buffer, DER_MAX_TBS);
        if (X509GetDeviceCertTBS(&leaf_ctx, &LEAF_CERT_TBS_DATA, &leaf_id_pub, (uint8_t*)&riot_info->ca_root_pub, sizeof(riot_info->ca_root_pub)) != 0)
```
I believe this is suppose to be used to build the x509 chain, so each leaf should be signed by the one above correct? so 1st leaf signed by ca_root_pub 2nd by the leaf pub we just created etc... ? So in this case since we would be signing via the cert above, wouldn't these be NULL, 0?

Reply:
@DMattoon PTAL